lua/twoslash-queries/init.lua

```diff
@@ -1,172 +1,343 @@
+  if M.config.multi_line == true then
+    local selected_lines = vim.list_slice(converted, limits[1], limits[2])
+    return selected_lines
+  end
+  local joined_string = table.concat(converted, "", limits[1], limits[2])
+  local escaped = string.gsub(joined_string, "  ", " ")
+  return string.sub(escaped, 1, 120)
+end
+
+local get_indent = function(line_num)
+  local indentexpr = vim.bo.indentexpr
+  if indentexpr ~= "" then
+    vim.v.lnum = line_num
+    local expr_indent_tbl = vim.api.nvim_exec2("echo " .. indentexpr, { output = true })
+    local expr_indent_str = expr_indent_tbl.output
+    local expr_indent = tonumber(expr_indent_str)
+    return expr_indent
+  end
+  local prev_nonblank = vim.fn.prevnonblank(line_num - 1)
+  local prev_nonblank_indent = vim.fn.indent(prev_nonblank)
+  return prev_nonblank_indent
+end
+
+M.add_query = function(pos)
+  local line, col = unpack(pos)
+  local indent = math.max(get_indent(line), get_indent(line + 1))
+  local two_slash_string = string.rep(" ", indent) .. "//"
+  two_slash_string = two_slash_string .. string.rep(" ", col - #two_slash_string) .. "^?"
+  vim.api.nvim_buf_set_lines(0, line, line, false, { two_slash_string })
 end
 
 M.remove_queries = function()
-	vim.api.nvim_buf_clear_namespace(0, virtual_types_ns, 0, -1)
+  vim.api.nvim_buf_clear_namespace(0, virtual_types_ns, 0, -1)
+
+  local lines = vim.api.nvim_buf_get_lines(0, 0, -1, false)
+  local removed_lines = 0
 
-	local lines = vim.api.nvim_buf_get_lines(0, 0, -1, false)
-	local regex = vim.regex(query_regex)
+  for index, line in pairs(lines) do
+    local match = query_regex:match_str(line)
 
-	local removed_lines = 0
+    if match then
+      vim.api.nvim_buf_set_lines(0, index - 1 - removed_lines, index - removed_lines, true, {})
+      removed_lines = removed_lines + 1
+    end
+  end
+end
 
-	for index, line in pairs(lines) do
-		local match = regex:match_str(line)
+---@alias line integer
+---@alias cache_item {column:integer,text:string,extmark:integer,clear:boolean,target:string}
+---@type Map<bufnr, Map<line, cache_item>>
+local cache = {}
 
-		if match then
-			vim.api.nvim_buf_set_lines(0, index - 1 - removed_lines, index - removed_lines, true, {})
-			removed_lines = removed_lines + 1
-		end
-	end
+local clear_cache = function(buffer_nr)
+  if cache[buffer_nr] then
+    for _, line in pairs(cache[buffer_nr]) do
+      line.clear = true
+    end
+  end
 end
 
-local get_types = function(client, buffer_nr)
-	vim.api.nvim_buf_clear_namespace(buffer_nr, virtual_types_ns, 0, -1)
+local update_hover_text = function(client, buffer_nr, line, column, cb)
+  local target = vim.api.nvim_buf_get_lines(buffer_nr, line, line + 1, false)[1]
+  local position = { line = line - 2, character = column - 1 }
+  if not vim.api.nvim_buf_is_valid(buffer_nr) then
+    cb()
+    return
+  end
+  local params = { textDocument = vim.lsp.util.make_text_document_params(buffer_nr), position = position }
+  if not client then
+    cb()
+    return
```

Comment:
we can join these conditions

Reply:
Hey @b0o Thank you for your awesome work on the project!  It took me a bit to digest all the changes, but that's probably on me 😂. I tested locally and everything works fine. I've left a couple of tiny comments.

After addressing these comments we can merge it as is for now and we can work on polishing it up after. I feel like we can improve the readability of the code but overall I think that Your improvements are solid!

Reply:
Thanks for reviewing! I will go ahead and clean things up a bit, I agree the readability can use improvement.